test/common/stats/thread_local_store_test.cc

```diff
@@ -80,6 +80,63 @@ class StatsThreadLocalStoreTest : public testing::Test, public RawStatDataAlloca
     store_->addSink(sink_);
   }
 
+  void validateMerge(std::string hist_name, std::vector<uint64_t> h1_cumulative_values,
+                     std::vector<uint64_t> h2_cumulative_values,
+                     std::vector<uint64_t> h1_interval_values,
+                     std::vector<uint64_t> h2_interval_values, bool single_histogram = false) {
+
+    std::shared_ptr<std::atomic<bool>> merge_called = std::make_shared<std::atomic<bool>>(false);
+    store_->mergeHistograms([merge_called]() -> void { *merge_called = true; });
+
+    EXPECT_TRUE(*merge_called);
+
+    std::list<HistogramSharedPtr> histogram_list = store_->histograms();
+    if (single_histogram) {
+      EXPECT_EQ(histogram_list.size(), 1);
+    } else {
+      EXPECT_EQ(histogram_list.size(), 2);
+    }
+    histogram_t* hist1_cumulative = hist_alloc();
+    for (uint64_t value : h1_cumulative_values) {
+      hist_insert_intscale(hist1_cumulative, value, 0, 1);
+    }
+
+    histogram_t* hist2_cumulative = hist_alloc();
+    for (uint64_t value : h2_cumulative_values) {
+      hist_insert_intscale(hist2_cumulative, value, 0, 1);
+    }
+
+    histogram_t* hist1_interval = hist_alloc();
+    for (uint64_t value : h1_interval_values) {
+      hist_insert_intscale(hist1_interval, value, 0, 1);
+    }
+
+    histogram_t* hist2_interval = hist_alloc();
+    for (uint64_t value : h2_interval_values) {
+      hist_insert_intscale(hist2_interval, value, 0, 1);
+    }
+
+    HistogramStatisticsImpl h1_cumulative_statistics(hist1_cumulative);
+    HistogramStatisticsImpl h2_cumulative_statistics(hist2_cumulative);
+    HistogramStatisticsImpl h1_interval_statistics(hist1_interval);
+    HistogramStatisticsImpl h2_interval_statistics(hist2_interval);
+
+    for (const Stats::HistogramSharedPtr& histogram : histogram_list) {
+      if (histogram->name().find(hist_name) != std::string::npos) {
+        EXPECT_EQ(histogram->cumulativeStatistics().summary(), h1_cumulative_statistics.summary());
+        // EXPECT_EQ(histogram->intervalStatistics().summary(), h1_interval_statistics.summary());
```

Comment:
i will enable this check based on @postwait reply on the issue I asked him.

Reply:
I can't find this communication.  Can you send this issue to me directly? Or file an issue on libcircllhist?

Reply:
@postwait Here is the issue
looks like found another corner case issue or may be i am misunderstanding.

Consider the code

histogram_t* main_thread_interval_hist = hist_alloc();
histogram_t* per_thread_interval_hist = hist_alloc();
hist_insert_intscale(per_thread_interval_hist, 1, 0, 1);
double in[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
double out[9];
hist_approx_quantile(main_thread_interval_hist, in, 9,out);
histogram_t* hist_array[1];
hist_array[0] = per_thread_interval_hist;
hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
hist_clear(per_thread_interval_hist);
//hist_free(per_thread_interval_hist);
//per_thread_interval_hist = hist_alloc();
hist_array[0] = per_thread_interval_hist;
hist_insert_intscale(per_thread_interval_hist, 2, 0, 1);
//hist_clear(main_thread_interval_hist);
//hist_free(main_thread_interval_hist);
main_thread_interval_hist = hist_alloc();
hist_accumulate(main_thread_interval_hist, hist_array, ARRAY_SIZE(hist_array));
hist_approx_quantile(main_thread_interval_hist, in, 9,out);
std::cout<<"after accumulations :"<<out[0]<<"==> \n";

histogram_t* direct_hist = hist_alloc();
hist_insert_intscale(direct_hist, 2, 0, 1);
hist_approx_quantile(direct_hist, in, 9,out);
std::cout<<"direct :"<<out[0]<<"==> \n";

I expect direct and after accumulations to give the same output. But it gives 1 in first bin (rest of the bins are same) for accumulation case and 2 in the direct case.

If I change the code to free and hist_alloc it works correctly in both the cases. Is this is a corner case issue with hist_clear?

Reply:
Sorry confusing example of allocating without freeing.  Hard to follow the example.  I think you're right... There's an issue with empty bins.

Reply:
Fixed upstream.

Reply:
@postwait thanks a lot. Updated the lib and it works great.